packages/libs/deer/json/src/lib.rs

```diff
@@ -387,47 +366,91 @@ impl<'a, 'de> deer::Deserializer<'de> for Deserializer<'a> {
     {
         try_deserialize!(match self {
             Value::Object(map) => visitor.visit_object(ObjectAccess::new(map, self.context)),
-            else => Error(schema: ObjectReflection)
+            else => Error
         })
     }
 }
 
 #[must_use]
 struct ArrayAccess<'a> {
+    dirty: bool,
     length: usize,
+
     inner: IntoIter<Value>,
     context: &'a Context,
+
+    remaining: Option<usize>,
 }
 
 impl<'a> ArrayAccess<'a> {
     fn new(array: Vec<Value>, context: &'a Context) -> Self {
         Self {
+            dirty: false,
             length: array.len(),
+
             inner: array.into_iter(),
             context,
+
+            remaining: None,
         }
     }
 }
 
 impl<'a, 'de> deer::ArrayAccess<'de> for ArrayAccess<'a> {
-    fn next<T>(&mut self) -> Result<Option<T>, ArrayAccessError>
+    fn set_bounded(&mut self, length: usize) -> Result<(), ArrayAccessError> {
+        if self.dirty {
+            return Err(
+                Report::new(SetBoundedError::Dirty.into_error()).change_context(ArrayAccessError)
+            );
+        }
+
+        if self.remaining.is_some() {
+            return Err(
+                Report::new(SetBoundedError::CalledMultipleTimes.into_error())
+                    .change_context(ArrayAccessError),
+            );
+        }
+
+        self.remaining = Some(length);
+
+        Ok(())
+    }
```

Comment:
I wonder why we want to fail on `set_bounded` in JSON at all. There are formats, which require the bound to be set, but JSON does not.

Reply:
`set_bounded` is agnostic to the `*Access` and would be set in the visitor. It enables the following pattern:

```
Expected: [Maybe<u32>; 5]
Given: [1, 2, 3]
Returned: [Maybe::Some(1), Maybe::Some(2), Maybe::Some(3), Maybe::None, Maybe::None]
```

Basically: `T::deserialize` is guaranteed to be called 5 times, no more and no less, and enables `visit_none` for items that do not exit. This enables default values, filler values, etc. I hope that makes sense, but to guarantee that it is precisely `n` items, we need to ensure that `set_bounded` is called at the start and only once.